### PR TITLE
Date Input

### PR DESCRIPTION
**Pull request by jd239**
## Summary of the changes
Date Input build.
 
Includes:
- Unit, E2E, Visual Regression and a11y automated
- Web component and react stories
- Updates to github action

## Related issue
#334 

## Checklist
- [x] I have added relevant unit and visual regression tests.
- [x] I have manually accessibility tested any changes, if relevant.
- [x] I have ensured any changes match the Figma component library.

### CODE REVIEW

**Comment by GCHQ-Developer-530** [comment]
I found a bit of a weird bug where the date input focuses if you click anywhere alongside it. Other than that, it looks really good 😊

https://github.com/mi6/ic-ui-kit/assets/111882034/da7a8338-d03a-4959-8d15-e313bd4fddc6

File: packages/web-components/src/components/ic-date-input/ic-date-input.css [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+}
+
+.year-input,
+.year-input.fit-to-value {
+  width: 37px;
+}
+
+.fit-to-value {
+  width: 20px;
+}
```
Comment (GCHQ-Developer-530) [inline packages/web-components/src/components/ic-date-input/ic-date-input.css:43]: Can these px values be converted to rem or variables?

File: packages/web-components/src/components/ic-date-input/ic-date-input.tsx [hunk 2]
```diff
@@ -0,0 +1,1132 @@
...
+
+  /**
+   * The label for the date input.
+   */
+  @Prop() label!: string;
+
+  /**
+   * If `true`, the input will require a value.
+   */
+  @Prop() required: boolean = false;
```
Comment (GCHQ-Developer-847) [inline packages/web-components/src/components/ic-date-input/ic-date-input.tsx:89]: A few of these props are missing the question mark to show that they're optional

**Review by GCHQ-Developer-847 — CHANGES_REQUESTED** [review]
Noticed a few things, but looking great overall!

Love the screen reader functionality in particular when using arrow keys to change the values - looks really good to me 👏

File: packages/web-components/src/components/ic-date-input/ic-date-input.tsx [hunk 3]
```diff
@@ -0,0 +1,1132 @@
...
+    // Highlight / select all characters in input for easy replacement
+    const input = event.target as HTMLInputElement;
+
+    input.select();
+    this.isDateSetFromKeyboardEvent = true;
+
+    this.setInputPreventBasedOnSelection(input);
+  };
+
+  private setInputPreventBasedOnSelection = (input: HTMLInputElement) => {
```
[the remark this answers is not in the document]
Comment (jd239) [reply packages/web-components/src/components/ic-date-input/ic-date-input.tsx:326]: In FF, if I add a value to day, (cursor moves to month), click back on day and then try and change the value, nothing happens. PreventInput was still set to true so this function is to check that the value matches the relevant day/month/year state and then set preventInput to false so the value can be updated. 

However, after you have said I've broken preventInput, fixing that might mean this is no longer required.

File: packages/web-components/src/components/ic-date-input/ic-date-input.tsx [hunk 4]
```diff
@@ -0,0 +1,1132 @@
...
+    }
+  };
+
+  // Set value of each input to value of pasted date part
+  // Allows formats separated by '-' or '/'
+  private handlePaste = (event: ClipboardEvent) => {
+    event.preventDefault();
+
+    const pastedValue = event.clipboardData.getData("Text");
+    //TODO Does this need to consider a string which includes Zulu time?
```
[the remark this answers is not in the document]
Comment (jd239) [reply packages/web-components/src/components/ic-date-input/ic-date-input.tsx:423]: Will create a separate ticket for this

File: packages/web-components/src/components/ic-date-input/ic-date-input.tsx (hunk 3 above)
[the remark this answers is not in the document]
Comment (jd239) [reply packages/web-components/src/components/ic-date-input/ic-date-input.tsx]: After further investigation, `window.getSelection.toString` does not work in FF. This method was created to as a work around which will work in both but another session of Googling provide a more reliable alternative (use selectionStart and selectionEnd). However, this involved changing the inputs from number to text. I have added regex to capture anything that is not number and swallow the keypress. All unit tests and e2e tests pass so think it's okay

File: packages/web-components/src/components/ic-date-input/ic-date-input.tsx
[the remark this answers is not in the document]
Comment (jd239) [reply packages/web-components/src/components/ic-date-input/ic-date-input.tsx:1]: This has been fixed.

**Comment by gd2910** [comment]
Components should now follow the [Stencil style guide](https://stenciljs.com/docs/style-guide)

**Comment by jd239** [comment]
> Components should now follow the [Stencil style guide](https://stenciljs.com/docs/style-guide)

Done :)

File: packages/web-components/src/components/ic-date-input/ic-date-input.tsx [hunk 5]
```diff
@@ -0,0 +1,1140 @@
...
+
+@Component({
+  tag: "ic-date-input",
+  styleUrl: "ic-date-input.css",
+  shadow: {
+    delegatesFocus: true,
+  },
+})
+export class DateInput {
+  @Element() el: HTMLIcDateInputElement;
```
Comment (gd2910) [inline packages/web-components/src/components/ic-date-input/ic-date-input.tsx:37]: Element should be inbetween private variables and states :)

**Comment by gd2910** [comment]
In the story 'with value, after changing the value and logging, the mm/yy values change but not the dd value, possibly just a SB story nuance?

**Comment by GCHQ-Developer-741** [comment]
> I found a bit of a weird bug where the date input focuses if you click anywhere alongside it. Other than that, it looks really good blush
> 
>  Screen.Recording.2023-06-19.at.13.38.18.mov

Sorry if I've missed discussion of this issue, but I've noticed this on other components in our design system. One that caught me out was the ic-switch component which I accidentally triggered doing the same thing in the video.

**Comment by jd239** [comment]
> In the story 'with value, after changing the value and logging, the mm/yy values change but not the dd value, possibly just a SB story nuance?

It was a problem with the component not update the relevant states.

**Comment by jd239** [comment]
> 

Was initially going to set the :host to inline-flex to fix this but then thought this will need to be added to all form controls. The issue with that is, switching from block to inline could break a lot applications which do not have the input wrapped in a blocked div so left it out. 

Will create a ticket to review for the next major version.

**Review by gd2910 — DISMISSED** [review]

**Comment by GCHQ-Developer-530** [comment]
I only seem to be able to paste a date into the input when I right click and select paste, it doesn't do anything when I use a keyboard shortcut. Is this meant to be the case or should I be able to use both methods?

**Review by gd2910 — APPROVED** [review]

**Comment by jd239** [comment]
Enhancement to paste functionality:
- Before it was possible to paste more than two characters into the day input. This has now been restricted to only two.
- It is now possible to paste individual 'numeric' values into the inputs. For example, it is now possible to copy and paste '12' into the month input.
- Validation has been set to prevent invalid values into relevant inputs for example, it is not possible to paste '13' into the month input. A validation message below will appear which says, 'Please paste a valid day/month'. This does not affect year as any number can be placed in there
- Moving to the next or previous input while the paste validation has appeared will remove the message.

**Comment by DBD324** [comment]
Couple of small potential improvements:
- The error message when attempting to paste an invalid value into a single filed is great. For example, pasting '13' into a month field. However, if I try to paste '123' in the field, then no message is displayed and the paste fails. It would be great to provide the same validation message and treat this as an invalid input in the same way. 
- When pasting a full date string '25/12/2017' we can use '/' and '-' as the separator keys. It would be great to also include the '.' symbol as a separtor allowing dates formatted '25.12.2017' to be pasted in the same way. 
- It would be great to allow the selection of the whole date when using the 'cmd+a' key press, and then allow the deletion of the whole date with one press of the delete key. 
- On the month element, should we allow a user to enter a letter corresponding to a month? For example, entering 'S' would result in an '09' entered into the month element. Entering 'J' would put the first month beginning with J - January - 01. Entering 'JU' would put the first month beginning with JU - June - 06. Entering JUL would then allow July - 07. 
- The allowable character entries for the above month text behaviour would be as follows:
-- JA
--F
---MAR
--AP
--MAY
--JUN
--JUL
--AU
--S
--O
--N
--D

**Review by DBD324 — CHANGES_REQUESTED** [review]
Have added comment with a few changes requested.

**Comment by jd239** [comment]
> Couple of small potential improvements:
> 
> * The error message when attempting to paste an invalid value into a single filed is great. For example, pasting '13' into a month field. However, if I try to paste '123' in the field, then no message is displayed and the paste fails. It would be great to provide the same validation message and treat this as an invalid input in the same way.
> * When pasting a full date string '25/12/2017' we can use '/' and '-' as the separator keys. It would be great to also include the '.' symbol as a separtor allowing dates formatted '25.12.2017' to be pasted in the same way.

These have been added.


> * It would be great to allow the selection of the whole date when using the 'cmd+a' key press, and then allow the deletion of the whole date with one press of the delete key.

Based on the three inputs which we use to generate the day/month/year, this does not seem possible as all three inputs would need to be focus and selected at the same time. 

> * On the month element, should we allow a user to enter a letter corresponding to a month? For example, entering 'S' would result in an '09' entered into the month element. Entering 'J' would put the first month beginning with J - January - 01. Entering 'JU' would put the first month beginning with JU - June - 06. Entering JUL would then allow July - 07.
> * The allowable character entries for the above month text behaviour would be as follows:
>   -- JA
>   --F
>   ---MAR
>   --AP
>   --MAY
>   --JUN
>   --JUL
>   --AU
>   --S
>   --O
>   --N
>   --D

This will be added to a new ticket rather than added to this one. The date input specifies 2 characters (which would be assumed as numbers instead of text) for the MM. Having the ability to type text in seems like a nice to have rather than a must have.

**Review by GCHQ-Developer-530 — APPROVED** [review]

**Pull request merged**